libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Env.hs

```diff
@@ -0,0 +1,154 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+-- | Provides variables (V era t), and mappings of them to objects of type 't'
+module Test.Cardano.Ledger.Constrained.Env (
+  V (..),
+  Field (..),
+  AnyF (..),
+  vToField,
+  fieldToV,
+  Env (..),
+  Payload (..),
+  emptyEnv,
+  findVar,
+  storeVar,
+  P (..),
+  bulkStore,
+  Dyn (..),
+  Name (..),
+  Access (..),
+) where
+
+import Data.List (intercalate)
+import qualified Data.List as List
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (isJust)
+import Data.Universe (Shaped (..))
+import Lens.Micro
+import Test.Cardano.Ledger.Constrained.Monad (Dyn (..), Typed (..), failT)
+import Test.Cardano.Ledger.Constrained.TypeRep
+
+-- ================================================================
+-- V
+
+-- | A proto variable. May or may not contain a Lens (encoded as Access)
+data V era t where V :: String -> Rep era t -> Access era s t -> V era t
+
+data Access era s t where
+  Yes :: forall s t era. Rep era s -> Lens' s t -> Access era s t
+  No :: Access era s t
+
+instance Show (V era t) where
+  show (V nm rep _) = nm ++ " :: " ++ show rep
+
+-- ===========================================================
+-- Name
+
+-- | An existentially quantified (V era t), hiding the 't'
+--   Usefull because unlike (V era t), it has both Eq and Ord instances
+data Name era where Name :: V era t -> Name era
+
+instance Show (Name era) where
+  show (Name (V n _ _)) = n
+
+-- | Does not satisfy extensionality
+instance Eq (Name era) where
+  Name (V n1 rep1 _) == Name (V n2 rep2 _) = n1 == n2 && isJust (testEql rep1 rep2)
+
+instance Ord (Name era) where
+  compare v1@(Name (V n1 rep1 _)) v2@(Name (V n2 rep2 _)) =
+    if v1 == v2
+      then EQ
+      else case compare n1 n2 of
+        LT -> LT
+        GT -> GT
+        EQ -> compareRep rep1 rep2
```

Comment:
That is actually pretty clever. I had to go look at the monoid instance for Ordering, to understand why it works.
It works because  (<>) on Ordering is left-biased (except for EQ).